routing/router.go

```diff
@@ -1749,36 +1750,97 @@ func outgoingFromIncoming(incomingAmt lnwire.MilliSatoshi,
 	// In case the inbound fee rate is less than -1 (-100%), we fail to
 	// compute the outbound amount and return the incoming amount. This also
 	// protects against zero division later.
-	m := (PPM*PPM + Ri*PPM + Ro*PPM + Ro*Ri) / (PPM * PPM)
-	if m <= 0 {
+	// m := (PPM*PPM + Ri*PPM + Ro*PPM + Ro*Ri) / (PPM * PPM)
+
+	// Note: this could maybe fit into an int64.
+	m1 := new(big.Int).Mul(PPM, PPM)
+	m2 := new(big.Int).Mul(Ri, PPM)
+	m3 := new(big.Int).Mul(Ro, PPM)
+	m4 := new(big.Int).Mul(Ro, Ri)
+
+	m := big.NewInt(0)
+	m.Add(m, m1)
+	m.Add(m, m2)
+	m.Add(m, m3)
+	m.Add(m, m4)
+
+	// Note: since we compare to 0, we can multiply by PPM * PPM to avoid
+	// division.
+	if m.Int64() <= 0 {
 		return incomingAmt
 	}
 
 	// In order to decide if the total fee is negative, we apply the fee
 	// to the *incoming* amount as mentioned before.
-	testAmtF := A + Bo + Bi + (PPM*(A*Ri+A*Ro+Ro*Ri)+A*Ri*Ro)/(PPM*PPM)
+	// testAmtF := A + Bo + Bi + (PPM*(A*Ri+A*Ro+Ro*Ri)+A*Ri*Ro)/(PPM*PPM)
+
+	// Note: this could maybe fit into an int64.
+	t1 := new(big.Int).Mul(A, Ri)
+	t2 := new(big.Int).Mul(A, Ro)
+	t3 := new(big.Int).Mul(Ro, Ri)
+
+	t4 := big.NewInt(0)
+	t4.Add(t4, t1)
+	t4.Add(t4, t2)
+	t4.Add(t4, t3)
+
+	t6 := new(big.Int).Mul(PPM, t4)
+
+	t7 := new(big.Int).Mul(A, Ri)
+	t7.Mul(t7, Ro)
+
+	num := new(big.Int).Add(t6, t7)
+	denom := new(big.Int).Mul(PPM, PPM)
+	fraction := new(big.Int).Div(num, denom)
+
+	// Sum up.
+	testAmt := big.NewInt(0)
+	testAmt.Add(testAmt, A)
+	testAmt.Add(testAmt, Bo)
+	testAmt.Add(testAmt, Bi)
+	testAmt.Add(testAmt, fraction)
 
 	// Protect against negative values for the integer cast to Msat.
-	if testAmtF < 0 {
+	if testAmt.Int64() < 0 {
 		return incomingAmt
 	}
 
 	// This is the second case where we have to compute the outgoing amount.
-	if lnwire.MilliSatoshi(testAmtF) > incomingAmt {
-		// Compute the exact outgoing amount as a float.
-		numerator := PPM * (PPM*(A-Bo-Bi) - Bo*Ri)
-		denominator := PPM*(PPM+Ri+Ro) + Ri*Ro
-		out := numerator / denominator
-
-		// We round to PPM precision.
-		rounded := math.Round(out*PPM) / PPM
+	if lnwire.MilliSatoshi(testAmt.Int64()) > incomingAmt {
+		// Compute the outgoing amount in terms of big.Int.
+		// numerator := PPM * (PPM*(A-Bo-Bi) - Bo*Ri)
+		// denominator := PPM*(PPM+Ri+Ro) + Ri*Ro
```

Comment:
I know that golang generally doesn't do this for whatever reason, but I would always put spaces between the operators and their operands.

Reply:
right, I just commented the auto-formatted lines, imo it's best to read if multiplication/division doesn't have spaces, but addition/subtraction has, change it that way